src/charm.py

```diff
@@ -0,0 +1,576 @@
+
+    @staticmethod
+    def _render_bessd_config_file(
+        upf_hostname: str,
+        upf_mode: str,
+        access_interface_name: str,
+        core_interface_name: str,
+        dnn: str,
+        pod_share_path: str,
+    ) -> str:
+        """Renders the configuration file for the 5G UPF service.
+
+        Args:
+            upf_hostname: UPF hostname
+            upf_mode: UPF mode
+            access_interface_name: Access network interface name
+            core_interface_name: Core network interface name
+            dnn: Data Network Name (DNN)
+            pod_share_path: pod_share path
+        """
+        jinja2_environment = Environment(loader=FileSystemLoader("src/templates/"))
+        template = jinja2_environment.get_template(f"{CONFIG_FILE_NAME}.j2")
+        content = template.render(
+            upf_hostname=upf_hostname,
+            mode=upf_mode,
+            access_interface_name=access_interface_name,
+            core_interface_name=core_interface_name,
+            dnn=dnn,
+            pod_share_path=pod_share_path,
+        )
+        return content
+
+    @property
+    def _upf_hostname(self) -> str:
+        return f"{self.model.app.name}.{self.model.name}.svc.cluster.local"
+
+    def _bessd_config_file_is_written(self) -> bool:
+        """Returns whether the bessd config file was written to the workload container.
+
+        Returns:
+            bool: Whether the bessd config file was written
+        """
+        if not self._bessd_container.exists(
+            path=f"{BESSD_CONTAINER_CONFIG_PATH}/{CONFIG_FILE_NAME}"
+        ):
+            return False
+        return True
+
+    def _bessd_config_file_content_matches(self, content: str) -> bool:
+        """Returns whether the bessd config file content matches the provided content.
+
+        Returns:
+            bool: Whether the bessd config file content matches
+        """
+        existing_content = self._bessd_container.pull(
+            path=f"{BESSD_CONTAINER_CONFIG_PATH}/{CONFIG_FILE_NAME}"
+        )
+        if existing_content.read() != content:
+            return False
+        return True
+
+    def _configure(self, event: EventBase) -> None:
+        """Handles events for configuring workloads.
+
+        Args:
+            event: EventBase
+        """
+        if invalid_configs := self._get_invalid_configs():
+            self.unit.status = BlockedStatus(
+                f"The following configurations are not valid: {invalid_configs}"
+            )
+            return
+        if not self._kubernetes_multus.multus_is_configured():
+            self.unit.status = WaitingStatus("Waiting for statefulset to be patched")
+            event.defer()
+            return
+        if not self._bessd_container.can_connect():
+            self.unit.status = WaitingStatus(
+                "Waiting to be able to connect to the `bessd` container"
+            )
+            event.defer()
```

Comment:
I see a lot of `defer` are back, wouldn't this be called again when we receive `PebbleReady`?

I assume this is related to the issue you had with crashes and I am curious to understand a bit more.

Reply:
Ok, thank you!